src/Bootstrappers/MailTenancyBootstrapper.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Stancl\Tenancy\Bootstrappers;
+
+use Illuminate\Config\Repository;
+use Stancl\Tenancy\Contracts\TenancyBootstrapper;
+use Stancl\Tenancy\Contracts\Tenant;
+
+class MailTenancyBootstrapper implements TenancyBootstrapper
+{
+    protected array $originalConfig = [];
+
+    /**
+     * Tenant properties to be mapped to config (similarly to the TenantConfig feature).
+     *
+     * For example:
+     * [
+     *     'config.key.name' => 'tenant_property',
+     * ]
+     */
+    public static array $credentialsMap = [
+        'mail.mailers.smtp.transport' => 'smtp_transport',
+        'mail.mailers.smtp.host' => 'smtp_host',
+        'mail.mailers.smtp.port' => 'smtp_port',
+        'mail.mailers.smtp.encryption' => 'smtp_encryption',
+        'mail.mailers.smtp.username' => 'smtp_username',
+        'mail.mailers.smtp.password' => 'smtp_password',
+        'mail.mailers.smtp.timeout' => 'smtp_timeout',
+        'mail.mailers.smtp.local_domain' => 'smtp_local_domain',
+    ];
```

Comment:
I'm looking at this again and I'm thinking that the previous approach could be a bit better for scenarios where you're setting completely new properties (e.g. because you're using a different mailer which has different properties). There setting `'aws_key' => 'mail.mailers.ses.whatever'` would make more sense.

**However**, I think that the scenario where you'd want to customize the properties easily, e.g. `$credentialsMap['mail.mailers.smtp.username'] = 'tenant_mail_username'`, makes more sense to prioritize.

So neither approach is perfect, but this one seems the most reasonable.

Another thing: I would simplify this array. We likely don't want to set things like `timeout` individually for each tenant. What happens if the property doesn't exist on the tenant and this bootstrapper is enabled? Does it simply get skipped (due to line 43)?

And I'm also thinking that it doesn't make sense to make this default to smtp since many people use different things. Maybe we could add static methods that would set the "preset", e.g. one for smtp, one for AWS SES etc.

Reply:
> What happens if the property doesn't exist on the tenant and this bootstrapper is enabled? Does it simply get skipped (due to line 43)?

If the property is `null` on the tenant, we skip setting the config, so the config value remains the same instead of being mapped to the tenant property (meaning if the tenant doesn't have the property, the default config value, e.g. the one set in a config file will be used).

Maybe checking if the value is `null` isn't right – someone might like to override a config value to make it `null`, but because we're skipping `null` values, the default wouldn't get overridden.

Reply:
> And I'm also thinking that it doesn't make sense to make this default to smtp since many people use different things. Maybe we could add static methods that would set the "preset", e.g. one for smtp, one for AWS SES etc.

Yeah, that makes sense.

Reply:
> Maybe checking if the value is null isn't right – someone might like to override a config value to make it null, but because we're skipping null values, the default wouldn't get overridden.

Check if `hasAttribute()` (or something like that) behaves differently than `getAttribute($foo) !== null`. Ideally the former would return true even on *defined* null. And return false only when the attribute isn't present in $attributes at all.

Reply:
I couldn't find a method similar to `hasAttribute()`, but it looks like `array_key_exists($attributeName, $tenant->getAttributes())` works.

Reply:
> We likely don't want to set things like timeout individually for each tenant

I agree with this. I made the SMTP preset have only `username`, `password`, `host`, and `port`. These looked like the ones that'd get mapped commonly (not sure about host and port, but I saw someone on Discord trying to map these, so I though these might be common).

> Maybe we could add static methods that would set the "preset", e.g. one for smtp, one for AWS SES etc.

[I updated the code](https://github.com/archtechx/tenancy/pull/989/files#diff-421166a13b8e50c082716efce511eec8587457ce0cac287118d5de829b4a4980R23-R46), could you take a look at it? I think instead of having a method per mailer, I'd like a big array with the presets which would look like this:
```php
public static $mapPresets = [
    'smtp' => [
        'mail.mailers.smtp.host' => 'smtp_host',
        'mail.mailers.smtp.port' => 'smtp_port',
        'mail.mailers.smtp.username' => 'smtp_username',
        'mail.mailers.smtp.password' => 'smtp_password',
    ],
    'ses' => [
        ...
    ]
]
```
Then, we could delete the `pickMapPreset` method, and instead of using it, we could just do `static::$mapPresets[$config->get('mail.default')] ?? []` in the constructor.

Reply:
I think this:

```php
public static function smtpPreset(): void
{
    $this->setMapping([
        'mail.mailers.smtp.host' => 'smtp_host',
        'mail.mailers.smtp.port' => 'smtp_port',
        'mail.mailers.smtp.username' => 'smtp_username',
        'mail.mailers.smtp.password' => 'smtp_password',
    ]);
}
```
would be better (with `setMapping()` being a simple method that sets the values; or maybe you can just directly modify the mapping as a property).

That way we don't make assumptions about the default mailer, and if people want to extend it to change the preset, they can easily do that.

But maybe a static property has its benefits too. That way people could just do `::$mapPresets['smtp'] = $theirPreset` without having to extend anything. So I guess the current approach is good, I'd just make it a bit less dependent on the default mailer. Maybe add an option to specify the mailer (probably some other static property).

Reply:
> So I guess the current approach is good, I'd just make it a bit less dependent on the default mailer. Maybe add an option to specify the mailer

I added the static property https://github.com/archtechx/tenancy/pull/989/commits/55aa49b18e7c2b227751a0e66bd0c0d35111cfb4

Reply:
I wrote:
> But maybe a static property has its benefits too. That way people could just do ::$mapPresets['smtp'] = $theirPreset without having to extend anything. So I guess the current approach is good, I'd just make it a bit less dependent on the default mailer

And it seems like we're using the method approach now. Do you have some reason for preferring that over a property?

Reply:
I prefer the property, I didn't change the method approach because you wrote "So I guess the current approach is good, I'd just make it a bit less dependent on the default mailer". I see you probably meant the approach I was writing about [here](https://github.com/archtechx/tenancy/pull/989#discussion_r1011849029)

Reply:
By current approach I meant the `$mapPresets` property, yeah. Didn't realize you changed it to a method before I described the method approach.

Reply:
Updated the code to use the static property now https://github.com/archtechx/tenancy/pull/989/commits/a0ad69a95cbd68bc49d46966d7623f64927a8d2f